Added support for --no-color flag in verbose and mini reporters

ThomasBem opened the pull request:
Started working on fixing #843. So far I have managed to disable color output for both mini and verbose reporters by setting the --no-color flag. 

Now I need to disable color output also if you try console.log with colors inside a test. What I am a little stuck on at the moment is this spawning of child processes. Thats a part of the code base I have yet to understand.

Any pointers on where to start digging? :)

novemberborn commented on lib/cli.js:
```diff
@@ -124,9 +125,14 @@ exports.run = () => {
 	} else if (cli.flags.verbose || isCi) {
-		reporter = verboseReporter();
+		reporter = verboseReporter({
+			disableColors: cli.flags.color !== undefined
```
Without looking into `meow`, I'm surprised this isn't a boolean. Does it become `false` if `--no-color` is set? Comparing against `false` seems better than comparing against `undefined`.

sindresorhus replied:
@novemberborn By default, if the user doesn't specify either `--color` or `--no-color`, it's not set. 

If we add the following, it will be `true` if not specified, but then we wouldn't know when it's not specified and should be auto-detected:

```js
const cli = meow(``, {
	boolean: ['color'],
	default: {
		color: true
	}
});
```

sindresorhus replied:
I would go with `color` instead of `disableColors` though. Double-negatives are not very good.

ThomasBem replied:
I will double check this. Was stuck on this point for a while trying out different things and reading up on meow and minimist. Got it working on undefined, but maybe I missed something.

ThomasBem replied:
@sindresorhus I tried to do what you suggested with default in in meow and that works for my code, 
`
 {
	color: true
}
`
but I don`t understand whats going on with what`s currently in default and how to make it work together with the change you recommended. Tried a few different things, but I keep getting a failing test in cli.js.

https://github.com/avajs/ava/blob/master/lib/cli.js#L75

novemberborn commented on lib/reporters/mini.js:
```diff
@@ -28,6 +28,13 @@ function MiniReporter(options) {
 
+	if (this.options.disableColors) {
+		chalk.enabled = false;
+		Object.keys(colors).forEach(function (key) {
```
I'd personally prefer a `for…of` loop here.

ThomasBem replied:
I just re-used the existing code that in other places did this but with "=true". But can make changes if you think that`s better.

https://github.com/ThomasBem/ava/blob/7c00c13844da6ab116676027b6482a57cf0f102a/lib/reporters/verbose.js#L9:L11

https://github.com/ThomasBem/ava/blob/7c00c13844da6ab116676027b6482a57cf0f102a/lib/reporters/mini.js#L14:L16

sindresorhus replied:
Yes, go with for-of whenever possible.

novemberborn commented:
> Now I need to disable color output also if you try console.log with colors inside a test

That seems unnecessary. If you want to log colors then go ahead. It's just that one should be able to disable AVA's color output.

ThomasBem commented:
@sindresorhus commented this in one of the PR`s that was working on this issue previously.

> It also needs to pass the --no-color flag down to the forked process. If I add console.log(require('chalk').red('colorzzzz')); in a test file, it will still log colors even with $ ava --no-colors. I would expect that to prevent all colors.

https://github.com/avajs/ava/pull/1104

novemberborn commented:
@ThomasBem I guess I disagree with @sindresorhus then 😄 

@sindresorhus why do you think `ava --no-color` should disable colors not coming from AVA itself?

sindresorhus commented:
> why do you think ava --no-color should disable colors not coming from AVA itself?

Because that's how it would work normally. That's how I designed Chalk. AVA is special since we spawn child processes, but the expected behavior should be preserved, but passing down the flag.

novemberborn commented:
Ah, hadn't realized `chalk` takes a `--no-color` argument.

We've previously argued against forwarding command line arguments to the workers though.

sindresorhus commented:
> We've previously argued against forwarding command line arguments to the workers though.

Yes, and we're not gonna start doing it with arbitrary flags, but I would say it's worth making an exception for this one as it matches expected outcome.

ThomasBem commented:
I have updated the PR with the suggested changes regarding the CLI flag and the for...of loops. 

I need some help with keeping the existing "default: conf" in cli.js while also adding the new stuff @sindresorhus mentioned. I could not get it to work together with the new { color: true} so I just removed it for now and its causing a test to fail.

I don`t understand what this conf thing is and what its doing in default in the first place. It makes no sense to me :P

I can understand adding { color: true } saying that color should be true if nothing else is set. But just "conf" makes no sense to me.

I also need some support in where to start with passing the flag down to the child processes. That`s one of the things I have yet to understand with how Ava works :)

novemberborn commented:
> I don`t understand what this conf thing is and what its doing in default in the first place. It makes no sense to me :P

This may be relevant: https://github.com/avajs/ava/issues/1046#issuecomment-275071873

> I could not get it to work together with the new { color: true} so I just removed it for now and its causing a test to fail.

The test verifies that config in `package.json` overrides the default behavior. Currently the config is passed through the CLI parser (`meow`) using `default`, except that you removed that behavior :wink:

If you do `Object.assign({ color: true }, conf)` instead you should get the behavior you want, and a passing test. I think this is another argument for my suggestion in https://github.com/avajs/ava/issues/1046#issuecomment-275071873.

> I also need some support in where to start with passing the flag down to the child processes. That`s one of the things I have yet to understand with how Ava works :)

See here: https://github.com/avajs/ava/blob/872d2edb938ab1104b304f25aa586833b9875030/lib/fork.js#L39

`opts` should already contain the `colors` value, you'll just have to translate it to the correct flag.

ThomasBem commented:
Thanks for the help @novemberborn :)
 
I did what you said for the default, and that worked fine. Think I _sort_ of understand whats going on with that now.

I also made something that works for passing the color flag down to the forked process.

I am unsure of how / if I can test that small piece of code though. Any suggestions?

forresst commented on lib/cli.js:
```diff
@@ -38,6 +38,7 @@ exports.run = () => {
 		  --verbose, -v           Enable verbose output
 		  --no-cache              Disable the transpiler cache
 		  --no-power-assert       Disable Power Assert
+		  --no-color              Disable color output
```
This flag should be added to readme.md (like -no-power-assert for example) in [CLI](https://github.com/avajs/ava/blob/master/readme.md#cli) paragraph

ThomasBem replied:
Good catch! Did not even think about that.

ThomasBem replied:
Should be good now :)

forresst commented on readme.md:
```diff
@@ -159,6 +159,7 @@ $ ava --help
     --verbose, -v           Enable verbose output
     --no-cache              Disable the transpiler cache
     --no-power-assert       Disable Power Assert
+	--no-color              Disable color output
```
Please, fix indentation

ThomasBem replied:
Hmm, ill take a look. It does not look that way for me in IDE or when looking at the readme in my branch. I see even Travis failed on some tests in node7 suddenly, when all i did was updated readme...

ThomasBem commented:
Looks better now? I used IntelliJ instead of VSCode to edit the markdown file.

ThomasBem commented:
Something going on with Travis? First time i edited the readme.md file, travis failed on node 7. 

Now I fixed a small indentation mistake and travis fails on node 4? Even different tests failing.

ThomasBem commented:
Seems like things are working much better today. So probably some hiccup with Travis yesterday.

sindresorhus commented on test/reporters/mini.js:
```diff
@@ -671,3 +671,26 @@ test('results when hasExclusive is enabled, but there are multiple remaining tes
 	t.end();
 });
 
+test('Result when no-color flag is set', function (t) {
```
Result => result

sindresorhus commented on test/reporters/verbose.js:
```diff
@@ -571,6 +571,31 @@ test('results when hasExclusive is enabled, but there are multiple remaining tes
 	t.end();
 });
 
+test('Result when no-color flag is set', function (t) {
```
Result => result

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -11,9 +11,9 @@ var repeating = require('repeating');
+for (let key of Object.keys(colors)) {
 	colors[key].enabled = true;
-});
+}
```
Why are we setting this both here and in the constructor? Wouldn't it make more sense to only set it in the constructor?

Something like this:

```js
chalk.enabled = this.options.color;
for (const key of Object.keys(colors)) {
	colors[key].enabled = this.options.color;
}
```

sindresorhus replied:
@Qix- This is needed because we assign shortcuts to the colors in [`lib/colors.js`](https://github.com/avajs/ava/blob/9616dde582e5333c3528cdd217c37e5b41798c9a/lib/colors.js) (context https://github.com/avajs/ava/pull/1104#issue-187486491) Any way we could make the above loop not nessecary? Maybe if we made `.enabled` property on the prototype?

ThomasBem replied:
You mean adding a MiniReporter.prototype.enabled = function () {} that would enable all the colors?

sindresorhus replied:
@ThomasBem Ignore this ;) Unrelated to this PR, but rather to [`chalk`](https://github.com/chalk/chalk).

ThomasBem replied:
Ah, oki :) I will commit with updates based on your review, but without doing anything regarding the loop.

I did a quick try where I changed how we create mini and verbose reporters in their test setup. Making sure to include {colors: true} as an options. That way we can use your suggestion:
```
chalk.enabled = this.options.color;
for (const key of Object.keys(colors)) {
	colors[key].enabled = this.options.color;
}
```

It works for the mini and verbose tests themselves, but it caused this test to fail because chalk and colors did not seem to work as intended, https://github.com/avajs/ava/blob/master/test/cli.js#L73

sindresorhus replied:
> but it caused this test to fail because chalk and colors did not seem to work as intended, https://github.com/avajs/ava/blob/master/test/cli.js#L73

I think that's as expected as we used to force Chalk in child processes and we no longer do. I guess we can just remove the `colors.error()` call from that test.

Qix- replied:
I'm a little lost, could you explain a bit more?

Qix- replied:
Nvm, Sindre explained it to me. See chalk/chalk#142.

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -28,6 +28,13 @@ function MiniReporter(options) {
 
+	if (this.options.color === false) {
+		chalk.enabled = false;
+		for (let key of Object.keys(colors)) {
```
`let` => `const`

sindresorhus commented:
@ThomasBem When you push new changes, can you push them as additional commits instead of squashing? Makes it easier to see what changed. We squash on merge anyways.

ThomasBem commented:
Sure, I can do that :)

novemberborn commented on lib/fork.js:
```diff
@@ -36,7 +36,9 @@ module.exports = (file, opts, execArgv) => {
 	}, opts);
 
-	const ps = childProcess.fork(path.join(__dirname, 'test-worker.js'), [JSON.stringify(opts)], {
+	var args = [JSON.stringify(opts), opts.color ? '--color' : '--no-color'];
```
Please use `const`.

(@sindresorhus is this something `xo` should enforce?)

sindresorhus replied:
I'll add the `esnext: true` option when https://github.com/avajs/ava/pull/1154 lands (hopefully in the next few days), as it's the last piece of non-ES2015ified code.

ThomasBem replied:
Sorry, C# habbit :p

novemberborn commented on lib/babel-config.js:
```diff
@@ -14,9 +12,9 @@ function validate(conf) {
+		let message = figures.cross;
 		message += ' Unexpected Babel configuration for AVA. ';
-		message += 'See ' + chalk.underline('https://github.com/avajs/ava#es2015-support') + ' for allowed values.';
+		message += 'See https://github.com/avajs/ava#es2015-support for allowed values.';
```
Why this change? (I must admit I didn't read all of the back and forth earlier though… forgive me if I overlooked the answer)

sindresorhus replied:
See: https://github.com/avajs/ava/pull/1198#discussion_r98352347

novemberborn replied:
@sindresorhus you suggested to remove the color comparison from the test, but this removes it from the actual AVA code.

Why do the changes in this PR require colors to be removed elsewhere? Should we then also remove colors from https://github.com/avajs/ava/blob/master/lib/cli.js#L100:L108?

ThomasBem replied:
I removed it from the actual output, since we removed it from test. But if the problem only exits when running tests, not real Ava. Then we can put it back in the implementation.

ThomasBem replied:
I did some testing and it seems like this is only a problem in test? Is this because execCli spawns a new process and the chalk.enabled and colors.enabled stuff is not passed on?

As far as I can tell none of the changes I did changed anything related to that behaviour? So why this did work before?

Since it seems to actually work when running Ava properly, we can always just remove the color from expected in the test? At least we catch that the error is thrown with the correct text. Even though colors and underline will be missing.

sindresorhus replied:
Oh, I didn't notice this removes it from the actual code. In https://github.com/avajs/ava/pull/1198#discussion_r98352347, I was talking about removing the chalk call from the test, not elsewhere.

> Is this because execCli spawns a new process and the chalk.enabled and colors.enabled stuff is not passed on?

Not exactly. Colors used to be forced "on", now it's auto-detected and colors are automatically off when spawned.

novemberborn commented on lib/reporters/verbose.js:
```diff
@@ -6,13 +6,16 @@ var plur = require('plur');
+		return new VerboseReporter(options);
+	}
+
+	this.options = Object.assign({}, options);
```
Assigning `this.options` seems unnecessary since the constructor didn't receive `options` before. Nothing else is using `this.options`.

Qix- commented:
The github mobile site doesn't allow inline comments, so apologies for this being out of line.

Foe the child workera, why don't we just utilize support color's new `FORCE_COLOR` environment variable instead of pushing down arbitrary flags?

sindresorhus commented:
@Qix- `FORCE_COLOR` is for users that can't pass flags. We control the child process, so we can pass flags. I don't really see the problem with that or any benefit in using `FORCE_COLOR`?

Qix- commented:
The parent AVA process is passed `--colors`, etc. Then it sets the environment variable `FORCE_COLOR` for child processes to _force_ them to match whatever color level the main (parent) process is running at. No need to do any messy flag juggling or whatever.

Removes the need to determine which flags to pass down. _This is what the environment was made for._

novemberborn commented:
@Qix- I think your explanation makes sense, but I'll defer to @sindresorhus who I'm sure has got a better understanding of this area than me.

sindresorhus commented:
> The parent AVA process is passed --colors, etc. Then it sets the environment variable FORCE_COLOR for child processes to force them to match whatever color level the main (parent) process is running at.

I still don't see how that's any easier. It's the same amount of work to pass down `--color`/`--no-color` based on those flags being set for AVA, as it's passing `FORCE_COLOR=0`/`FORCE_COLOR=1` based on AVA receiving `--color`/`--no-color`. The benefit of passing the flags is that we still respect `FORCE_COLOR` being passed to AVA by the user.

> No need to do any messy flag juggling or whatever. No need to determine which flags to pass down. No need to update flags in AVA in the future if supports-color changes how it handles flags. This is the perfect use case.

I don't do what-ifs. We work with how it works now. In the future the environment variable might be renamed, who knows.

novemberborn commented:
> The benefit of passing the flags is that we still respect FORCE_COLOR being passed to AVA by the user.

👍

ThomasBem commented:
So I have reverted back to babel-config.js change I made, and instead passed down an additional --color flag in execCli where needed.

Tok me quite a while to get my work in after @vadimdemedes big Magic Assert PR. But I think I managed to get everything up-to-date. But please take a look :)

sindresorhus commented:
Great work @ThomasBem! Another high-quality pull request :)
